### PR TITLE
Better support for enum trees

### PR DESCRIPTION
**Pull request by soc**
Replace the approach of “detect some magic combination of flags to
trigger some action” by introducing an enum flag which makes the
semantics and the intentions of the code using it more explicit.

This basically involves wiring up the existing ACC_ENUM bit to a new
modifier flag and making sure it is set correctly when encountering
enums.

The existing enum tests files/pos/t5165 and files/pos/t2764 keep
working, showing that this hasn't introduced any obvious regressions.

Tests for the changes in Namer which prevent scalac from messing with
enum trees can't be added yet, because one currently can't define an
enum in Scala without the help of the macro paradise.
The intention is to add the @enum macro as a full test suite as soon
as one can depend on macro annotations.

### CODE REVIEW

**Comment by adriaanm** [comment]
Getting close!

```
fail - pos/t2764  [compilation failed]% scalac t2764/Ann.java t2764/Enum.java t2764/Use.scala
Use.scala:4: error: annotation argument needs to be a constant; found: Enum.VALUE
  @Ann(Enum.VALUE)
            ^
one error found

fail - pos/t5165  [compilation failed]% scalac t5165/TestAnnotation.java t5165/TestObject.scala t5165/TestTrait.scala
TestTrait.scala:2: error: annotation argument needs to be a constant; found: TestAnnotation.TestEnumOne.A
@TestAnnotation(one=TestAnnotation.TestEnumOne.A, two=TestAnnotation.TestEnumTwo.C, strVal="something")
                                               ^
TestTrait.scala:2: error: annotation argument needs to be a constant; found: TestAnnotation.TestEnumTwo.C
@TestAnnotation(one=TestAnnotation.TestEnumOne.A, two=TestAnnotation.TestEnumTwo.C, strVal="something")
                                                                                 ^
two errors found

```

File: src/reflect/scala/reflect/internal/Flags.scala [hunk 1]
```diff
@@ -119,6 +119,7 @@ class ModifierFlags {
   final val DEFAULTINIT   = 1L << 41      // symbol is initialized to the default value: used by -Xcheckinit
   final val ARTIFACT      = 1L << 46      // symbol should be ignored when typechecking; will be marked ACC_SYNTHETIC in bytecode
   final val DEFAULTMETHOD = 1L << 47      // symbol is a java default method
+  final val ENUM          = 1L << 48      // symbol is an enum
```
Comment (xeno-by) [inline src/reflect/scala/reflect/internal/Flags.scala:122]: You meant an enum class? What about enum fields? Do they also get this flag? Would be nice to have tests for that.

Comment (soc) [inline src/reflect/scala/reflect/internal/Flags.scala:122]: I'll add a better comment. Both enum classes ant enum constants are getting this flag.

**Comment by soc** [comment]
@adriaanm Kind of weird that it worked after the first commit, but failed in the second one. I'll investigate!

**Comment by soc** [comment]
Superseded by https://github.com/scala/scala/pull/3247.

**Pull request closed without merging**